workflow/validate/validate.go

```diff
@@ -932,20 +932,30 @@ func (ctx *templateValidationCtx) validateSteps(scope map[string]interface{}, tm
 			if err != nil {
 				return err
 			}
-		}
 
-		stepBytes, err := json.Marshal(stepGroup)
-		if err != nil {
-			return errors.InternalWrapError(err)
-		}
-		err = resolveAllVariables(scope, ctx.globalParams, string(stepBytes), workflowTemplateValidation)
-		if err != nil {
-			return errors.Errorf(errors.CodeBadRequest, "templates.%s.steps %s", tmpl.Name, err.Error())
-		}
+			stepBytes, err := json.Marshal(step)
+			if err != nil {
+				return errors.InternalWrapError(err)
+			}
+
+			stepScope := make(map[string]interface{})
+			for k, v := range scope {
+				stepScope[k] = v
+			}
+
+			if i := step.Inline; i != nil {
+				for _, p := range i.Inputs.Parameters {
+					stepScope["inputs.parameters."+p.Name] = placeholderGenerator.NextPlaceholder()
+				}
+			}
+
+			err = resolveAllVariables(stepScope, ctx.globalParams, string(stepBytes), workflowTemplateValidation)
+			if err != nil {
+				return errors.Errorf(errors.CodeBadRequest, "templates.%s.steps %s", tmpl.Name, err.Error())
+			}
 
-		for _, step := range stepGroup.Steps {
```

Comment:
Why do we need to remove this loop? will Code iterate all steps in stepGroup?

Reply:
It's not this loop that _needed_ to be removed, it was the serialization of the `stepGroup` that I removed a few lines above - the input parameters for the inline template aren't in the scope and so validation fails. They can't just be added to the overall scope or they would be visible to all steps in the group (AFAIK).

And then once that has been done you have two loops right next to each other looping over the same slice (`stepGroup.Steps`) - it seems to make sense that you should merge them then.